1 js/08 prototypes/04 prototype methods (#139)
javascript-tutorial/ru.javascript.info: 1-js/08-prototypes/04-prototype-methods/article.md

# Методы прототипов, объекты без свойства __proto__

В первой главе этого раздела мы упоминали, что существуют современные методы работы с прототипами.

> Свойство `__proto__` считается устаревшим и в некотором роде нежелательным (только в браузерной части стандарта JavaScript).

Reviewer: Свойство `__proto__` считается устаревшим и по стандарту должно поддерживаться только браузерами.

В первой главе этого раздела мы упоминали, что существуют современные методы работы с прототипами.

> Свойство `__proto__` считается устаревшим и по стандарту должно поддерживаться только браузерами.

Reviewer: , и по стандарту

```

Такой вызов создаёт точную копию объекта `obj`, включая все свойства: перечисляемые и не перечисляемые, свойства, геттеры/сеттеры для свойств -- и все это с правильным свойством `[[Prototype]]`.

## Краткая история

Если пересчитать все способы управления прототипом, то их будет много! И многие из них делают одно и тоже!

Почему так?

В силу исторических причин.

- Свойство `"prototype"` функции-конструктора существует с совсем давних времён.
- Позднее, в 2012 году, в стандарте появился метод `Object.create`. Это давало возможность создавать объекты с указанным прототипом, но не позволяло устанавливать/получать его. Тогда браузеры реализовали нестандартный аксессор `__proto__`, который позволил устанавливать/получать прототип в любой время.
- Позднее в 2015 году: в стандарт были добавлены `Object.setPrototypeOf` и `Object.getPrototypeOf,` заменяющие собой аксессор `__proto__`, который упоминается в Приложении Б стандарта, которое не обязательно к поддержке в не-браузерных окруженях. При этом де-факто `__proto__` всё ещё поддерживается везде.

В итоге сейчас у нас есть все эти способы для работы с прототипом.

> Почему же __proto__ был заменён на функции `getPrototypeOf/setPrototypeOf`? Читайте далее, чтобы узнать ответ.

Reviewer: тут нужны кавычки (обычные обратные) вокруг __proto__, может мы недопоняли друг друга? =)

alert(rabbit.jumps); // true
```

> У дескрипторов такой же формат как описано в главе <info:property-descriptors>.

Reviewer: Формат задания дескрипторов описан в главе <info:property-descriptors>.

> Мы можем использовать `Object.create` для клонирования объектов как более мощный способ чем копирование свойств с помощью цикла `for..in`:

Reviewer: Мы также можем использовать `Object.create` для "продвинутого" клонирования объекта, более мощного, чем копирование свойств в цикле `for..in`:

```js

> // полностью идентичный поверхностный клон obj

Reviewer: клон obj c тем же прототипом (с поверхностным копированием свойств)

let clone = Object.create(Object.getPrototypeOf(obj), Object.getOwnPropertyDescriptors(obj));
```

Такой вызов создаёт точную копию объекта `obj`, включая все свойства: перечисляемые и не перечисляемые, свойства, геттеры/сеттеры для свойств -- и все это с правильным свойством `[[Prototype]]`.

## Краткая история

> Если пересчитать все способы управления свойством `[[Prototype]]`, то их будет много! И многие из них делают одно и тоже!

Reviewer: Если пересчитать все способы управления прототипом, то их будет много! И многие из них делают одно и тоже!

Почему так?

В силу исторических причин.

> - Свойство `"prototype"` функции-конструктора существует ещё с очень давних времён.

Reviewer: Свойство `"prototype"` функции-конструктора существует с совсем давних времён.

> - Позднее в 2012 году: `Object.create` появился в стандарте. Это давало возможность создавать объекты с указанным прототипом, но не позволяло устанавливать/получать его. Тогда браузеры реализовали нестандартный аксессор `__proto__`, который позволил устанавливать/получать прототип в любой время.

Reviewer: Позднее, в 2012 году, в стандарте появился метод `Object.create`.

> - Позднее в 2015 году: в стандарт были добавлены `Object.setPrototypeOf` и `Object.getPrototypeOf`. Фактически свойство `__proto__` было реализовано везде, так что оно попало Приложение Б стандарта, которое было опционально для не браузерных окружений.

Reviewer: Позднее в 2015 году: в стандарт были добавлены `Object.setPrototypeOf` и `Object.getPrototypeOf`, заменяющие собой аксессор `__proto__`, который упоминается в Приложении Б стандарта, которое не обязательно к поддержке в не-браузерных окруженях. При этом де-факто `__proto__` всё ещё поддерживается везде.

> Теперь мы знаем, что у нас есть все эти способы в нашем распоряжении.

Reviewer: В итоге сейчас у нас есть все эти способы для работы с прототипом.

> Почему же `__proto__` был заменён на функции? Вопрос интересный, требующий от нас понимания, почему `__proto__` плох. Читайте далее, чтобы узнать ответ.

Reviewer: Почему же `__proto__` был заменён на функции `getPrototypeOf/setPrototypeOf`? Читайте далее, чтобы узнать ответ.

```warn header="Не переопределяйте `[[Prototype]]`, кроме случаев, когда скорость не важна"
Технически, мы можем установить/получить `[[Prototype]]` в любое время. Но обычно мы устанавливаем прототип только раз во время создания объекта, а после не меняем: `rabbit` наследует от `animal`, и это не изменится.

> И JavaScript движки хорошо оптимизированы для этого. Изменение прототипа "на лету" с помощью `Object.setPrototypeOf` или `obj.__proto__=` очень медленная операция, которая ломает внутренние оптимизации для операций доступа к свойствам объекта. Так что лучше избегать этого, кроме, тех ситуаций, когда вы знаете, что делаете, или скорость JavaScript для вас полностью не важна.

Reviewer 2: ```suggestion
И JavaScript движки хорошо оптимизированы для этого. Изменение прототипа "на лету" с помощью `Object.setPrototypeOf` или `obj.__proto__=` очень медленная операция, которая ломает внутренние оптимизации для операций доступа к свойствам объекта. Так что лучше избегайте этого, кроме тех случаев, когда вы знаете, что делаете, либо скорость JavaScript для вас не имеет никакого значения.
```

```

## "Очень пустой" объект

Как мы знаем, объекты можно использовать как ассоциативные массивы для хранения пар ключ/значение.

...Но если попробуем хранить *созданные пользователями* ключи (например, словари с пользовательским вводом), мы можем заметить интересный сбой: все ключи работают как ожидается, за исключением `"__proto__"`.

Посмотрите на пример:

```js run
let obj = {};

let key = prompt("What's the key?", "__proto__");
obj[key] = "some value";

alert(obj[key]); // [object Object], не "some value"!
```

Если пользователь введёт `__proto__`, присвоение проигнорируется!

И это не должно удивлять нас. Свойство `__proto__` особенное: оно должно быть либо объектом, либо `null`, а строка не может стать прототипом.

Но мы не *намеревались* реализовывать такое поведение, не так ли? Мы хотим хранить пары ключ/значение, и ключ с именем `"__proto__"` не был сохранён надлежащим образом. Так что это ошибка!

> В примере последствия не так ужасны. Но в других случаях прототип может быть изменён, и код может выполниться совершенно неожиданным неправильным способом.

Reviewer: Конкретно в этом примере последствия не так ужасны, но если мы присваиваем объектные значения, то прототип и в самом деле может быть изменен. В результате дальнейшее выполнение пойдет совершенно непредсказуемым образом.

Что хуже всего -- разработчики не задумываются о такой возможности совсем. Это делает такие ошибки сложным для отлавливания или даже превращает их в уязвимости, особенно когда JavaScript используется на сервере.

Неожиданные вещи могут случаться также при получении доступа к свойству `toString`, которое по умолчанию функция, и к другим встроенным свойствам.

Как же избежать проблемы?

Во-первых, мы можем переключиться на использование коллекции `Map`, и тогда все будет в порядке.

Но и `Object` может также хорошо подойти, потому что создатели языка уже давно продумали решение проблемы.

> Свойство `__proto__` это не свойство объекта, а свойство-аксессор для `Object.prototype`:

Reviewer: Свойство `__proto__` -- не обычное, а аксессор, заданный в `Object.prototype`:

![](object-prototype-2.png)

> Так что при чтении или установке `obj.__proto__`, вызывается соответствующий геттер/сеттер из прототипа `obj`, и именно он устанавливает/получает свойство `[[Prototype]]`.

Reviewer: лишняя запятая первая

> Как было сказано в начале этой обучающей секции: `__proto__` это способ доступа к свойству `[[Prototype]]`, это не само свойство `[[Prototype]]`.

Reviewer: Как было сказано в начале этой секции учебника, `__proto__`  --

alert(obj[key]); // "some value"
```

> `Object.create(null)` создаёт пустой объект без прототипа (свойство `[[Prototype]]` это `null`):

Reviewer: (`[[Prototype]]` будет `null`):

![](object-prototype-null.png)

> Таким образом не будет унаследованного геттера/сеттера для `__proto__`. Теперь это свойство обрабатывается как обычное свойство, и пример приведённый выше работает правильно.

Reviewer: и приведенный выше пример

> Мы можем назвать такой объект "очень пустым" или "чистым словарным объектом", потому что они ещё проще чем обычные объекты `{...}`.

Reviewer: "очень пустым" - везде плиз замени термин на "простейший"

alert(obj); // Error (no toString)
```

> ...Но обычно этого достаточно для ассоциативных массивов.

Reviewer: ...Но обычно это нормально для ассоциативных массивов.

> Пожалуйста, обратите внимание, что большая часть методов `Object.something(...)` связанных с объектами, таких как `Object.keys(obj)`, не находятся в прототипе, так что они продолжат работать для таких объектов:

Reviewer 2: ```suggestion
Пожалуйста, обратите внимание, что большая часть методов, связанных с объектами, имеют вид `Object.something(...)`. К примеру `Object.keys(obj)` не находятся в прототипе, так что они продолжат работать для таких объектов:
```

alert(Object.keys(chineseDictionary)); // hello,bye
```

## Итого

Современные способы установки и прямого доступа к прототипу это:

- [Object.create(proto[, descriptors])](mdn:js/Object/create) -- создаёт пустой объект со свойством `[[Prototype]]`, указанным как `proto` (может быть `null`), и необязательными дескрипторами свойств.

> - [Object.getPrototypeOf(obj)](mdn:js/Object.getPrototypeOf) -- возвращает свойство `[[Prototype]]` объекта `obj` (то же самое что и геттер `__proto__`).

Reviewer 2: ```suggestion
- [Object.getPrototypeOf(obj)](mdn:js/Object.getPrototypeOf) -- возвращает свойство `[[Prototype]]` объекта `obj` (то же самое, что и геттер `__proto__`).
```

- [Object.setPrototypeOf(obj, proto)](mdn:js/Object.setPrototypeOf) -- устанавливает свойство `[[Prototype]]` объекта `obj` как `proto` (то же самое что и сеттер `__proto__`).

> Встроенный геттер/сеттер __proto__ не безопасен, если мы хотим использовать *созданные пользователями* ключи в объекте. Как минимум потому, что пользователь может ввести "proto" как ключ, от чего может возникнуть ошибка. Если повезёт - последствия будут лёгкими, но обычно они непредсказуемы.

Reviewer: может ввести `__proto__` (забыты подчеркивания) как ключ

> Тогда мы можем использовать либо `Object.create(null)` для создания "очень пустого" объекта, либо использовать коллекцию `Map`.

Reviewer: простейшего

> Также, `Object.create` даёт нам лёгкий способ создать поверхностную копию объекты со всеми дескрипторами:

Reviewer 2: ```suggestion
Кроме этого, `Object.create` даёт нам лёгкий способ создать поверхностную копию объекта со всеми дескрипторами:
```

```js
let clone = Object.create(Object.getPrototypeOf(obj), Object.getOwnPropertyDescriptors(obj));
```

- [Object.keys(obj)](mdn:js/Object/keys) / [Object.values(obj)](mdn:js/Object/values) / [Object.entries(obj)](mdn:js/Object/entries) -- возвращают массив всех перечисляемых собственных строковых имён/значений/пар ключ-значение.
- [Object.getOwnPropertySymbols(obj)](mdn:js/Object/getOwnPropertySymbols) -- возвращает массив всех собственных символьных имён свойств.
- [Object.getOwnPropertyNames(obj)](mdn:js/Object/getOwnPropertyNames) -- возвращает массив всех собственных строковых имён свойств.
- [Reflect.ownKeys(obj)](mdn:js/Reflect/ownKeys) -- возвращает массив всех собственных имён свойств.
- [obj.hasOwnProperty(key)](mdn:js/Object/hasOwnProperty): возвращает `true`, если у `obj` есть собственное (не унаследованное) свойство с именем `key`.

> Мы также ясно указали, что  `__proto__` -- это геттер/сеттер для свойства `[[Prototype]]` и находится он в `Object.prototype`, как и другие методы.

Reviewer: Мы также ясно увидели, что

> Мы можем создавать объекты без прототипов с помощью `Object.create(null)`. Такие объекты можно использовать как "чистые словари", у них нет проблем с использованием строки `"__proto__"` как ключ.

Reviewer: в качестве ключа

> Все методы, которые возвращают свойства объектов (такие как `Object.keys` и другие), возвращают "собственные" свойства. Если мы хотим получить унаследованные, тогда мы можем использовать цикл `for..in`.

Reviewer 2: ```suggestion
Все методы, которые возвращают свойства объектов (такие как `Object.keys` и другие), возвращают "собственные" свойства. Если мы хотим получить унаследованные, тогда можем воспользоваться циклом `for..in`.
```
